File: android/leokit/src/main/kotlin/io/github/leoho0722/leokit/components/LKCard.kt
package io.github.leoho0722.leokit.components

import androidx.compose.foundation.BorderStroke
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import io.github.leoho0722.leokit.LKShapes
import io.github.leoho0722.leokit.LKTheme
import io.github.leoho0722.leokit.tokens.LKSize
import io.github.leoho0722.leokit.tokens.LKSpacing

/**
 * 卡片看起來浮起多高。
 *
 * 深色主題不用陰影表達高低，改用 bgElevated 那一階較亮的底色，所以深色下請用 [Flat]。
 */
public enum class LKCardElevation {
    /** 貼平在背景上，不畫陰影。 */
    Flat,

    /** 預設：貼著背景但有一點陰影。 */
    Resting,

    /** 明顯浮在其他內容之上，用於被拉出來強調的卡片。 */
    Raised,
}

/**
 * 把一組相關的內容收成一個看得出邊界的面。
 *
 * 卡片裡不要再放卡片，層層疊起來會讓人看不出哪一塊是重點。
 *
 * @param modifier 由呼叫端套在整張卡片上的修飾子
 * @param elevation 卡片看起來浮起多高
 * @param selected 卡片目前是否被選取；選取時邊框換成品牌色並加粗
 * @param flush 內容是否貼齊卡片邊緣；放整張圖或整列清單時設為 true
 * @param content 卡片裡要放的內容，以直排方式擺放
 */
@Composable
public fun LKCard(
    modifier: Modifier = Modifier,
    elevation: LKCardElevation = LKCardElevation.Resting,
    selected: Boolean = false,
    flush: Boolean = false,
    content: @Composable ColumnScope.() -> Unit,
) {
    val colors = LKTheme.colors
    val tonal: Dp = when (elevation) {
        LKCardElevation.Flat -> 0.dp
        LKCardElevation.Resting -> 1.dp
        LKCardElevation.Raised -> 2.dp
    }
    Card(
        modifier = modifier,
        shape = LKShapes.lg,
        colors = CardDefaults.cardColors(
            containerColor = colors.bgSurface,
            contentColor = colors.textPrimary,
        ),
        elevation = CardDefaults.cardElevation(defaultElevation = tonal),
        border = BorderStroke(
            width = if (selected) LKSize.borderWStrong else LKSize.hairline,
            color = if (selected) colors.brand else colors.borderSubtle,
        ),
    ) {
        Column(
            modifier = Modifier.padding(
                if (flush) PaddingValues(0.dp) else PaddingValues(LKSpacing.spacing16)
            ),
            content = content,
        )
    }
}

File: ios/Sources/LeoKit/Components/LKCard.swift
//
//  LKCard.swift
//  LeoKit
//
//  Created by Leo Ho on 2026/09/19.
//

import SwiftUI

/// 把一組相關的內容收成一個看得出邊界的面
///
/// - Note: 卡片裡不要再放卡片，層層疊起來會讓人看不出哪一塊是重點
public struct LKCard<Content: View>: View {

    // MARK: - Properties

    /// 目前是淺色還是深色外觀，深色下不畫陰影
    @Environment(\.colorScheme) private var colorScheme

    /// 卡片裡要放的內容
    private let content: Content

    /// 卡片看起來浮起多高
    private let elevation: LKCardElevation

    /// 內容是否貼齊卡片邊緣，放整張圖或整列清單時設為 true
    private let flush: Bool

    /// 卡片目前是否被選取，選取時邊框換成品牌色並加粗
    private let selected: Bool

    // MARK: - Init

    /// 建立一張卡片
    ///
    /// - Parameters:
    ///   - elevation: 卡片看起來浮起多高
    ///   - selected: 卡片目前是否被選取
    ///   - flush: 內容是否貼齊卡片邊緣
    ///   - content: 卡片裡要放的內容
    public init(
        elevation: LKCardElevation = .resting,
        selected: Bool = false,
        flush: Bool = false,
        @ViewBuilder content: () -> Content
    ) {
        self.elevation = elevation
        self.selected = selected
        self.flush = flush
        self.content = content()
    }

    // MARK: - Body

    /// 卡片的骨架：內容加上內距，再套上底色、外框與陰影
    public var body: some View {
        content
            .padding(contentPadding)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(LKColor.bgSurface, in: LKShape.lg)
            .overlay {
                border
            }
            .clipShape(LKShape.lg)
            .shadow(color: shadowColor, radius: shadowRadius, y: shadowOffset)
    }
}

// MARK: - Private Views

private extension LKCard {

    /// 卡片的外框，被選取時換成品牌色並加粗
    var border: some View {
        LKShape.lg
            .stroke(borderColor, lineWidth: borderWidth)
    }
}

// MARK: - Nested Types

extension LKCard {

    /// 卡片自己的版面數值，只放沒有對應 token 的尺寸
    ///
    /// - Note: 一律 computed 不用 stored：`Layout` 巢狀在泛型型別裡時
    ///   static stored property 不合法，八個元件統一寫法才不用每次判斷
    private enum Layout {

        // MARK: - Computed Properties

        /// 貼著背景時的陰影濃度
        static var restingShadowOpacity: Double { 0.06 }

        /// 明顯浮起時的陰影濃度
        static var raisedShadowOpacity: Double { 0.08 }
    }
}

// MARK: - Computed Properties

private extension LKCard {

    /// 內容四周的內距，貼齊邊緣時為 0
    var contentPadding: CGFloat {
        if flush {
            0
        } else {
            LKSpacing.spacing16
        }
    }

    /// 外框的顏色
    var borderColor: Color {
        if selected {
            LKColor.brand
        } else {
            LKColor.borderSubtle
        }
    }

    /// 外框的粗細
    var borderWidth: CGFloat {
        if selected {
            LKSize.borderWStrong
        } else {
            LKSize.hairline
        }
    }

    /// 陰影的顏色，深色外觀下完全透明等於不畫
    var shadowColor: Color {
        Color.black.opacity(shadowOpacity)
    }

    /// 陰影的濃度；深色外觀靠底色分層，所以一律為 0
    var shadowOpacity: Double {
        if colorScheme == .dark {
            0
        } else {
            switch elevation {
            case .flat:
                0
            case .resting:
                Layout.restingShadowOpacity
            case .raised:
                Layout.raisedShadowOpacity
            }
        }
    }

    /// 陰影的模糊半徑
    var shadowRadius: CGFloat {
        switch elevation {
        case .flat:
            0
        case .resting:
            1
        case .raised:
            3
        }
    }

    /// 陰影往下偏移多少
    var shadowOffset: CGFloat {
        switch elevation {
        case .flat:
            0
        case .resting:
            1
        case .raised:
            2
        }
    }
}

// MARK: - Preview

#Preview("三種高度") {
    VStack(spacing: LKSpacing.spacing16) {
        LKCard(elevation: .flat) {
            Text("貼平在背景上")
                .font(LKFont.body)
        }

        LKCard {
            Text("預設高度")
                .font(LKFont.body)
        }

        LKCard(elevation: .raised, selected: true) {
            Text("浮起並被選取")
                .font(LKFont.body)
        }
    }
    .padding(LKSpacing.spacing16)
    .background(LKColor.bgCanvas)
}
